parcels/fieldfilebuffer.py

```diff
@@ -0,0 +1,565 @@
+        lon0 = self.indices['lon'][0]
+        lon1 = self.indices['lon'][-1] + 1
+        if len(data.shape) == 3:
+            data = self.lib.concatenate((data[d0:d1 - 1, lat0:lat1, lon0:lon1],
+                                        da.zeros((1, lat1 - lat0, lon1 - lon0))), axis=0)
+        else:
+            if (type(ti) in [list, range]):
+                t0 = ti[0]
+                t1 = ti[-1] + 1
+                data = self.lib.concatenate((data[t0:t1, d0:d1 - 1, lat0:lat1, lon0:lon1],
+                                             self.lib.zeros((t1 - t0, 1, lat1 - lat0, lon1 - lon0))), axis=1)
+            else:
+                data = self.lib.concatenate((data[ti, d0:d1 - 1, lat0:lat1, lon0:lon1],
+                                             self.lib.zeros((1, lat1 - lat0, lon1 - lon0))), axis=0)
+
+        return data
+
+    def _apply_indices(self, data, ti):
+        if len(data.shape) == 2:
+            data = data[self.indices['lat'], self.indices['lon']]
+        elif len(data.shape) == 3:
+            if self._check_extend_depth(data, 0):
+                data = self._extend_depth_dimension(data, ti)
+            elif len(self.indices['depth']) > 1:
+                data = data[self.indices['depth'], self.indices['lat'], self.indices['lon']]
+            else:
+                data = data[ti, self.indices['lat'], self.indices['lon']]
+        else:
+            if self._check_extend_depth(data, 1):
+                data = self._extend_depth_dimension(data, ti)
+            else:
+                data = data[ti, self.indices['depth'], self.indices['lat'], self.indices['lon']]
+
+        return data
+
+    @property
+    def data(self):
+        return self.data_access()
+
+    def data_access(self):
+        data = self.dataset[self.name]
+        ti = range(data.shape[0]) if self.ti is None else self.ti
+        data = self._apply_indices(data, ti)
+        return np.array(data)
+
+    @property
+    def time(self):
+        return self.time_access()
+
+    def time_access(self):
+        if self.timestamp is not None:
+            return self.timestamp
+
+        if 'time' not in self.dimensions:
+            return np.array([None])
+
+        time_da = self.dataset[self.dimensions['time']]
+        convert_xarray_time_units(time_da, self.dimensions['time'])
+        time = np.array([time_da[self.dimensions['time']]]) if len(time_da.shape) == 0 else np.array(time_da[self.dimensions['time']])
+        if isinstance(time[0], datetime.datetime):
+            raise NotImplementedError('Parcels currently only parses dates ranging from 1678 AD to 2262 AD, which are stored by xarray as np.datetime64. If you need a wider date range, please open an Issue on the parcels github page.')
+        return time
+
+
+class DeferredNetcdfFileBuffer(NetcdfFileBuffer):
+    def __init__(self, *args, **kwargs):
+        super(DeferredNetcdfFileBuffer, self).__init__(*args, **kwargs)
+
+
+class DaskFileBuffer(NetcdfFileBuffer):
```

Comment:
not further reviewing this in-depth, as this will be adapted in another PR later - just looking if all required functions are there